File: app/src/main/java/muryshkin/alexey/pdd/Data/DataHolder.java
package muryshkin.alexey.pdd.Data;

import android.content.Context;
import android.content.res.AssetManager;
import android.support.v4.util.Pair;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by 123 on 7/7/2016.
 */
public class DataHolder {

    private static DataHolder dataHolder;

    public static DataHolder getDataHolder() {
       if (dataHolder == null)
           dataHolder = new DataHolder();
        return dataHolder;
    }

    public boolean isLoggedIn;

    public List<JSONObject> pdd;
    public List<JSONObject> signs;

    public List<Integer> answers;
    public List<Integer> correctAnswers;
    public int result;
    public int answered;

    public List<Test> bTests;
    public List<Integer> maxCorrect;
    public JSONObject test;

    public JSONObject chosenArticle;

    public DataHolder() {
        isLoggedIn = false;

        chosenArticle = null;

        maxCorrect = new ArrayList<>();
        bTests = new ArrayList<>();
        test = null;

        correctAnswers = new LinkedList<>();
        answers = new LinkedList<>();
        result = 0;
        answered = 0;

        pdd = new LinkedList<>();
        getData("pdd");

        signs = new LinkedList<>();
        getData("signs");
    }

    private void getData(String fileName) {

        String json;

        AssetManager assetManager = Const.context.getAssets();

        try {
            InputStream inputStream = assetManager.open(fileName + ".json", Const.context.MODE_PRIVATE);
            int size = inputStream.available();

            byte[] buffer = new byte[size];
            inputStream.read(buffer);

            inputStream.close();

            json = new String(buffer, "UTF-8");
            JSONArray data = new JSONArray(json);

            requestHelper(data, fileName);
        } catch (IOException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void requestHelper(JSONArray data, String fileName) {

        for (int i = 0; i < data.length(); i++)
            try {
                if (fileName.equals("pdd"))
                    pdd.add(data.getJSONObject(i));
                else if (fileName.equals("signs"))
                    signs.add(data.getJSONObject(i));
            } catch (JSONException e) {
                e.printStackTrace();
            }
    }
}

File: app/src/main/java/muryshkin/alexey/pdd/Data/DataHolder.kt
package muryshkin.alexey.pdd.Data

import android.content.Context
import android.content.Context.MODE_PRIVATE
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.util.*
import kotlin.text.Charsets.UTF_8

/**
 * Created by 123 on 7/7/2016.
 */
class DataHolder {

    private var trafficRulesData: MutableMap<String, List<JSONObject>> = HashMap()

    private var answers: List<Int> = ArrayList()
    private var correctAnswers: List<Int> = ArrayList()
    private var result: Int = 0
    private var answered: Int = 0

    private var bTests: List<Test> = ArrayList()
    private var maxCorrect: List<Int> = ArrayList()
    private var test: JSONObject? = null

    private var chosenArticle: JSONObject? = null

    init {
        storeData(
                TRAFFIC_RULES,
                retrieveAssetsData(TRAFFIC_RULES, FILE_EXTENSION_JSON))
        storeData(
                TRAFFIC_SIGNS,
                retrieveAssetsData(TRAFFIC_SIGNS, FILE_EXTENSION_JSON))
    }

    private fun retrieveAssetsData(fileName: String, fileExtension: String): String? {

        val context: Context
        val assetManager = context.assets

        try {
            val inputStream = assetManager.open(
                    "$fileName.$fileExtension", MODE_PRIVATE)
            val size = inputStream.available()
            val buffer = ByteArray(size)

            inputStream.read(buffer)
            inputStream.close()

            return String(buffer, UTF_8)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    private fun storeData(key: String, data: String?): Boolean {
        if (data == null) {
            return false
        }

        try {
            val jsonArray = JSONArray(data)
            val dataArray = ArrayList<JSONObject>()

            for (i in 0 until jsonArray.length()) {
                dataArray.add(jsonArray.getJSONObject(i))
            }

            trafficRulesData[key] = dataArray
            return true
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return false
    }
}

private var dataHolder: DataHolder? = null

fun getDataHolder(): DataHolder {
    if (dataHolder == null)
        dataHolder = DataHolder()
    return dataHolder!!
}
